komfydb/execution/aggregate.cc

```diff
@@ -0,0 +1,230 @@
+#include "komfydb/execution/aggregate.h"
+#include "komfydb/common/field.h"
+#include "komfydb/execution/aggregate_tuple.h"
+#include "komfydb/execution/aggregator.h"
+#include "komfydb/execution/op.h"
+#include "komfydb/storage/page_id.h"
+#include "komfydb/storage/record_id.h"
+#include "komfydb/utils/status_macros.h"
+
+#include "absl/status/status.h"
+#include "absl/status/statusor.h"
+
+namespace {
+
+typedef komfydb::execution::Aggregator::AggregateType AggregateType;
+using komfydb::common::Field;
+using komfydb::common::Type;
+using komfydb::execution::AggregateTuple;
+using komfydb::storage::PageId;
+using komfydb::storage::RecordId;
+
+/* TODO: there's too much field copying here */
+
+absl::Status FillGroupFields(Tuple& group, std::vector<int>& groupby_fields,
+                             std::vector<Type>& groupby_types, Record* record) {
+  for (int i = 0; i < groupby_fields.size(); i++) {
+    ASSIGN_OR_RETURN(Field * field, record->GetField(groupby_fields[i]));
+    switch (groupby_types[i].GetValue()) {
+      case Type::INT: {
+        std::unique_ptr<IntField> group_field =
+            std::make_unique<IntField>(*dynamic_cast<IntField*>(field));
+        RETURN_IF_ERROR(group.SetField(i, std::move(group_field)));
+        break;
+      }
+      case Type::STRING: {
+        std::unique_ptr<StringField> group_field =
+            std::make_unique<StringField>(*dynamic_cast<StringField*>(field));
+        RETURN_IF_ERROR(group.SetField(i, std::move(group_field)));
+        break;
+      }
+    }
+  }
+  return absl::OkStatus();
+}
+
+absl::Status UpdateGroup(AggregateTuple* group,
+                         std::vector<int>& aggregate_fields,
+                         std::vector<AggregateType>& aggregate_types,
+                         Record* record) {
+  for (int i = 0; i < aggregate_fields.size(); i++) {
+    ASSIGN_OR_RETURN(Field * group_field, group->GetField(i));
+    ASSIGN_OR_RETURN(Field * new_field, record->GetField(aggregate_fields[i]));
+    RETURN_IF_ERROR(group_field->ApplyAggregate(aggregate_types[i], new_field));
+  }
+  group->IncremetGroupSize();
+  return absl::OkStatus();
+}
+
+absl::Status InitializeGroup(AggregateTuple* group,
+                             std::vector<int>& aggregate_fields,
+                             std::vector<AggregateType>& aggregate_types,
+                             Record* record) {
+  for (int i = 0; i < aggregate_fields.size(); i++) {
+    ASSIGN_OR_RETURN(Field * field, record->GetField(aggregate_fields[i]));
+    switch (aggregate_types[i]) {
+      case AggregateType::COUNT: {
+        RETURN_IF_ERROR(group->SetField(i, std::make_unique<IntField>(1)));
```

Comment:
Why not 0?

Reply:
This is called when we ecounter first group record